--- ../src-base/minecraft/net/minecraft/block/state/pattern/BlockStateMatcher.java
+++ ../src-work/minecraft/net/minecraft/block/state/pattern/BlockStateMatcher.java
@@ -12,13 +12,7 @@
 
 public class BlockStateMatcher implements Predicate<IBlockState>
 {
-    public static final Predicate<IBlockState> ANY = new Predicate<IBlockState>()
-    {
-        public boolean apply(@Nullable IBlockState p_apply_1_)
-        {
-            return true;
-        }
-    };
+    public static final Predicate<IBlockState> ANY = p_apply_1_ -> true;
     private final BlockStateContainer blockstate;
     private final Map < IProperty<?>, Predicate<? >> propertyPredicates = Maps. < IProperty<?>, Predicate<? >> newHashMap();
 
