--- ../src-base/minecraft/org/bukkit/command/defaults/VersionCommand.java
+++ ../src-work/minecraft/org/bukkit/command/defaults/VersionCommand.java
@@ -44,8 +44,7 @@
 
         if (args.length == 0) {
             sender.sendMessage("This server is running " + Bukkit.getName() + " version " + Bukkit.getVersion() + " (Implementing API version " + Bukkit.getBukkitVersion() + ")");
-            sender.sendMessage(ChatColor.YELLOW + "This is a final build for 1.12.2. Please see https://www.spigotmc.org/ for details about upgrading.");
-            // sendVersion(sender);
+            sendVersion(sender);
         } else {
             StringBuilder name = new StringBuilder();
 
@@ -199,7 +198,6 @@
                     setVersionMessage("You are " + (cbVersions + spigotVersions) + " version(s) behind");
                 }
             }
-
         } else if (version.startsWith("git-Bukkit-")) {
             version = version.substring("git-Bukkit-".length());
             int cbVersions = getDistance("craftbukkit", version.substring(0, version.indexOf(' ')));
