--- ../src-base/minecraft/net/minecraftforge/fml/repackage/com/nothome/delta/DiffWriter.java
+++ ../src-work/minecraft/net/minecraftforge/fml/repackage/com/nothome/delta/DiffWriter.java
@@ -47,7 +47,6 @@
 	 * Flushes to output, e.g. any data added.
 	 */
     public void flush() throws IOException;
-
     /**
      * Closes this stream.
      * Note that {@link DiffWriter} will invoke this method at the end.
@@ -55,4 +54,3 @@
     @Override
     public void close() throws IOException;
 }
-
