--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/FMLLaunchHandler.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/FMLLaunchHandler.java
@@ -92,7 +92,7 @@
         FMLInjectionData.build(minecraftHome, classLoader);
         redirectStdOutputToLog();
         FMLLog.log.info("Forge Mod Loader version {}.{}.{}.{} for Minecraft {} loading", FMLInjectionData.major, FMLInjectionData.minor,
-                FMLInjectionData.rev, FMLInjectionData.build, FMLInjectionData.mccversion);
+            FMLInjectionData.rev, FMLInjectionData.build, FMLInjectionData.mccversion);
         FMLLog.log.info("Java is {}, version {}, running on {}:{}:{}, installed at {}", System.getProperty("java.vm.name"), System.getProperty("java.version"), System.getProperty("os.name"), System.getProperty("os.arch"), System.getProperty("os.version"), System.getProperty("java.home"));
         FMLLog.log.debug("Java classpath at launch is:");
         for (String path : System.getProperty("java.class.path").split(File.pathSeparator))
