--- ../src-base/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
+++ ../src-work/minecraft/net/minecraftforge/fml/relauncher/ServerLaunchWrapper.java
@@ -19,18 +19,21 @@
 
 package net.minecraftforge.fml.relauncher;
 
+import org.kettlemc.utils.DownloadServerFiles;
+
 import java.lang.reflect.Method;
 
-import org.apache.logging.log4j.LogManager;
-import org.apache.logging.log4j.core.LoggerContext;
-
 public class ServerLaunchWrapper {
 
+    public static long beginTime;
+
     /**
      * @param args
      */
-    public static void main(String[] args)
-    {
+    public static void main(String[] args) {
+        System.out.println("Starting Kettle");
+        DownloadServerFiles.downloadMinecraftServer();
+        DownloadServerFiles.downloadServerLibraries();
         new ServerLaunchWrapper().run(args);
     }
 
@@ -56,9 +59,9 @@
         {
             System.err.printf("We appear to be missing one or more essential library files.\n" +
             		"You will need to add them to your server before FML and Forge will run successfully.");
-            e.printStackTrace(System.err);
-            System.exit(1);
+
         }
+        beginTime = System.nanoTime();
 
         try
         {
