--- ../src-base/minecraft/net/minecraft/world/biome/Biome.java
+++ ../src-work/minecraft/net/minecraft/world/biome/Biome.java
@@ -49,6 +49,7 @@
 import net.minecraftforge.fml.relauncher.SideOnly;
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public abstract class Biome extends net.minecraftforge.registries.IForgeRegistryEntry.Impl<Biome>
 {
@@ -182,7 +183,7 @@
                 return this.spawnableCaveCreatureList;
             default:
                 // Forge: Return a non-empty list for non-vanilla EnumCreatureTypes
-                if (!this.modSpawnableLists.containsKey(creatureType)) this.modSpawnableLists.put(creatureType, Lists.<Biome.SpawnListEntry>newArrayList());
+                if (!this.modSpawnableLists.containsKey(creatureType)) this.modSpawnableLists.put(creatureType, Lists.<SpawnListEntry>newArrayList());
                 return this.modSpawnableLists.get(creatureType);
         }
     }
@@ -475,7 +476,8 @@
             return;
         }
 
-        world.setBlockState(pos, flower.state, 3);
+//        world.setBlockState(pos, flower.state, 3);
+        CraftEventFactory.handleBlockGrowEvent(world, pos.getX(), pos.getY(), pos.getZ(), flower.state.getBlock(), flower.state.getBlock().getMetaFromState(flower.state));
     }
 
     /* ========================================= FORGE END ======================================*/
