--- ../src-base/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
+++ ../src-work/minecraft/net/minecraft/network/rcon/RConConsoleSource.java
@@ -22,11 +22,20 @@
         return "Rcon";
     }
 
+    public void sendMessage(String message) {
+        this.buffer.append(message);
+    }
+
     public void sendMessage(ITextComponent component)
     {
         this.buffer.append(component.getUnformattedText());
     }
 
+    public boolean canUseCommand(int permLevel, String commandName, String perm)
+    {
+        return true;
+    }
+
     public boolean canUseCommand(int permLevel, String commandName)
     {
         return true;
