--- ../src-base/minecraft/net/minecraftforge/fml/common/Loader.java
+++ ../src-work/minecraft/net/minecraftforge/fml/common/Loader.java
@@ -97,6 +97,7 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParser;
+import org.kettlemc.api.ServerAPI;
 
 import javax.annotation.Nullable;
 
@@ -757,6 +758,7 @@
         modController.distributeStateMessage(LoaderState.AVAILABLE);
         GameData.freezeData();
         FMLLog.log.info("Forge Mod Loader has successfully loaded {} mod{}", mods.size(), mods.size() == 1 ? "" : "s");
+        ServerAPI.mods.put("mods", mods.size());
         progressBar.step("Completing Minecraft initialization");
     }
 
@@ -904,8 +906,10 @@
             }
         }
 
-        if (difference.size() > 0)
+        if (difference.size() > 0){
+            ServerAPI.modList.addAll(difference);
             FMLLog.log.info("Attempting connection with missing mods {} at {}", difference, side);
+        }
         return true;
     }
 
