--- ../src-base/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
+++ ../src-work/minecraft/net/minecraft/entity/item/EntityFireworkRocket.java
@@ -23,13 +23,14 @@
 import net.minecraft.world.World;
 import net.minecraftforge.fml.relauncher.Side;
 import net.minecraftforge.fml.relauncher.SideOnly;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class EntityFireworkRocket extends Entity
 {
-    private static final DataParameter<ItemStack> FIREWORK_ITEM = EntityDataManager.<ItemStack>createKey(EntityFireworkRocket.class, DataSerializers.ITEM_STACK);
+    public static final DataParameter<ItemStack> FIREWORK_ITEM = EntityDataManager.<ItemStack>createKey(EntityFireworkRocket.class, DataSerializers.ITEM_STACK);
     private static final DataParameter<Integer> BOOSTED_ENTITY_ID = EntityDataManager.<Integer>createKey(EntityFireworkRocket.class, DataSerializers.VARINT);
     private int fireworkAge;
-    private int lifetime;
+    public int lifetime;
     private EntityLivingBase boostedEntity;
 
     public EntityFireworkRocket(World worldIn)
@@ -187,8 +188,10 @@
 
         if (!this.world.isRemote && this.fireworkAge > this.lifetime)
         {
-            this.world.setEntityState(this, (byte)17);
-            this.dealExplosionDamage();
+            if (!org.bukkit.craftbukkit.event.CraftEventFactory.callFireworkExplodeEvent(this).isCancelled()) {
+                this.world.setEntityState(this, (byte) 17);
+                this.dealExplosionDamage();
+            }
             this.setDead();
         }
     }
@@ -209,7 +212,9 @@
         {
             if (this.boostedEntity != null)
             {
+                CraftEventFactory.entityDamage = this;
                 this.boostedEntity.attackEntityFrom(DamageSource.FIREWORKS, (float)(5 + nbttaglist.tagCount() * 2));
+                CraftEventFactory.entityDamage = null;
             }
 
             double d0 = 5.0D;
@@ -235,7 +240,9 @@
                     if (flag)
                     {
                         float f1 = f * (float)Math.sqrt((5.0D - (double)this.getDistance(entitylivingbase)) / 5.0D);
+                        CraftEventFactory.entityDamage = this;
                         entitylivingbase.attackEntityFrom(DamageSource.FIREWORKS, f1);
+                        CraftEventFactory.entityDamage = null;
                     }
                 }
             }
