--- ../src-base/minecraft/net/minecraftforge/fml/server/FMLServerHandler.java
+++ ../src-work/minecraft/net/minecraftforge/fml/server/FMLServerHandler.java
@@ -19,7 +19,11 @@
 
 package net.minecraftforge.fml.server;
 
-import java.io.*;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.io.InputStream;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Set;
