--- ../src-base/minecraft/net/minecraft/block/BlockPotato.java
+++ ../src-work/minecraft/net/minecraft/block/BlockPotato.java
@@ -27,13 +27,13 @@
     {
         super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
 
-        if (false && !worldIn.isRemote) //Forge: Moved to getDrops
-        {
-            if (this.isMaxAge(state) && worldIn.rand.nextInt(50) == 0)
-            {
-                spawnAsEntity(worldIn, pos, new ItemStack(Items.POISONOUS_POTATO));
-            }
-        }
+//        if (!worldIn.isRemote) //Forge: Moved to getDrops
+//        {
+//            if (this.isMaxAge(state) && worldIn.rand.nextInt(50) == 0)
+//            {
+//                spawnAsEntity(worldIn, pos, new ItemStack(Items.POISONOUS_POTATO));
+//            }
+//        }
     }
 
     public AxisAlignedBB getBoundingBox(IBlockState state, IBlockAccess source, BlockPos pos)
