--- ../src-base/minecraft/net/minecraft/block/BlockCrops.java
+++ ../src-work/minecraft/net/minecraft/block/BlockCrops.java
@@ -15,6 +15,7 @@
 import net.minecraft.util.math.MathHelper;
 import net.minecraft.world.IBlockAccess;
 import net.minecraft.world.World;
+import org.bukkit.craftbukkit.event.CraftEventFactory;
 
 public class BlockCrops extends BlockBush implements IGrowable
 {
@@ -81,7 +82,8 @@
 
                 if(net.minecraftforge.common.ForgeHooks.onCropsGrowPre(worldIn, pos, state, rand.nextInt((int)(25.0F / f) + 1) == 0))
                 {
-                    worldIn.setBlockState(pos, this.withAge(i + 1), 2);
+                    IBlockState data = this.withAge(i + 1);
+                    CraftEventFactory.handleBlockGrowEvent(worldIn, pos.getX(), pos.getY(), pos.getZ(), this, getMetaFromState(data));
                     net.minecraftforge.common.ForgeHooks.onCropsGrowPost(worldIn, pos, state, worldIn.getBlockState(pos));
                 }
             }
@@ -98,7 +100,8 @@
             i = j;
         }
 
-        worldIn.setBlockState(pos, this.withAge(i), 2);
+        IBlockState data = this.withAge(i);
+        CraftEventFactory.handleBlockGrowEvent(worldIn, pos.getX(), pos.getY(), pos.getZ(), this, getMetaFromState(data));
     }
 
     protected int getBonemealAgeIncrease(World worldIn)
@@ -202,23 +205,23 @@
     {
         super.dropBlockAsItemWithChance(worldIn, pos, state, chance, fortune);
 
-        if (false && !worldIn.isRemote) // Forge: NOP all this.
-        {
-            int i = this.getAge(state);
-
-            if (i >= this.getMaxAge())
-            {
-                int j = 3 + fortune;
-
-                for (int k = 0; k < j; ++k)
-                {
-                    if (worldIn.rand.nextInt(2 * this.getMaxAge()) <= i)
-                    {
-                        spawnAsEntity(worldIn, pos, new ItemStack(this.getSeed()));
-                    }
-                }
-            }
-        }
+//        if (!worldIn.isRemote) // Forge: NOP all this.
+//        {
+//            int i = this.getAge(state);
+//
+//            if (i >= this.getMaxAge())
+//            {
+//                int j = 3 + fortune;
+//
+//                for (int k = 0; k < j; ++k)
+//                {
+//                    if (worldIn.rand.nextInt(2 * this.getMaxAge()) <= i)
+//                    {
+//                        spawnAsEntity(worldIn, pos, new ItemStack(this.getSeed()));
+//                    }
+//                }
+//            }
+//        }
     }
 
     public Item getItemDropped(IBlockState state, Random rand, int fortune)
