--- ../src-base/minecraft/net/minecraft/block/BlockSlab.java
+++ ../src-work/minecraft/net/minecraft/block/BlockSlab.java
@@ -128,40 +128,40 @@
         {
             return false;
         }
-        else if (false) // Forge: Additional logic breaks doesSideBlockRendering and is no longer useful.
-        {
-            IBlockState iblockstate = blockAccess.getBlockState(pos.offset(side));
-            boolean flag = isHalfSlab(iblockstate) && iblockstate.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP;
-            boolean flag1 = isHalfSlab(blockState) && blockState.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP;
-
-            if (flag1)
-            {
-                if (side == EnumFacing.DOWN)
-                {
-                    return true;
-                }
-                else if (side == EnumFacing.UP && super.shouldSideBeRendered(blockState, blockAccess, pos, side))
-                {
-                    return true;
-                }
-                else
-                {
-                    return !isHalfSlab(iblockstate) || !flag;
-                }
-            }
-            else if (side == EnumFacing.UP)
-            {
-                return true;
-            }
-            else if (side == EnumFacing.DOWN && super.shouldSideBeRendered(blockState, blockAccess, pos, side))
-            {
-                return true;
-            }
-            else
-            {
-                return !isHalfSlab(iblockstate) || flag;
-            }
-        }
+//        else if (false) // Forge: Additional logic breaks doesSideBlockRendering and is no longer useful.
+//        {
+//            IBlockState iblockstate = blockAccess.getBlockState(pos.offset(side));
+//            boolean flag = isHalfSlab(iblockstate) && iblockstate.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP;
+//            boolean flag1 = isHalfSlab(blockState) && blockState.getValue(HALF) == BlockSlab.EnumBlockHalf.TOP;
+//
+//            if (flag1)
+//            {
+//                if (side == EnumFacing.DOWN)
+//                {
+//                    return true;
+//                }
+//                else if (side == EnumFacing.UP && super.shouldSideBeRendered(blockState, blockAccess, pos, side))
+//                {
+//                    return true;
+//                }
+//                else
+//                {
+//                    return !isHalfSlab(iblockstate) || !flag;
+//                }
+//            }
+//            else if (side == EnumFacing.UP)
+//            {
+//                return true;
+//            }
+//            else if (side == EnumFacing.DOWN && super.shouldSideBeRendered(blockState, blockAccess, pos, side))
+//            {
+//                return true;
+//            }
+//            else
+//            {
+//                return !isHalfSlab(iblockstate) || flag;
+//            }
+//        }
         return super.shouldSideBeRendered(blockState, blockAccess, pos, side);
     }
 
