--- ../src-base/minecraft/net/minecraft/block/BlockShulkerBox.java
+++ ../src-work/minecraft/net/minecraft/block/BlockShulkerBox.java
@@ -44,7 +44,7 @@
 public class BlockShulkerBox extends BlockContainer
 {
     public static final PropertyEnum<EnumFacing> FACING = PropertyDirection.create("facing");
-    private final EnumDyeColor color;
+    public final EnumDyeColor color;
 
     public BlockShulkerBox(EnumDyeColor colorIn)
     {
@@ -160,9 +160,32 @@
         }
     }
 
-    public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune)
-    {
+    // CraftBukkit start - override to prevent duplication when dropping
+    public void dropBlockAsItemWithChance(World worldIn, BlockPos pos, IBlockState state, float chance, int fortune) {
+        TileEntity tileentity = worldIn.getTileEntity(pos);
+
+        if (tileentity instanceof TileEntityShulkerBox) {
+            TileEntityShulkerBox tileentityshulkerbox = (TileEntityShulkerBox) tileentity;
+
+            if (!tileentityshulkerbox.isCleared() && tileentityshulkerbox.shouldDrop()) {
+                ItemStack itemstack = new ItemStack(Item.getItemFromBlock(this));
+                NBTTagCompound nbttagcompound = new NBTTagCompound();
+                NBTTagCompound nbttagcompound1 = new NBTTagCompound();
+
+                nbttagcompound.setTag("BlockEntityTag", ((TileEntityShulkerBox) tileentity).saveToNbt(nbttagcompound1));
+                itemstack.setTagCompound(nbttagcompound);
+                if (tileentityshulkerbox.hasCustomName()) {
+                    itemstack.setStackDisplayName(tileentityshulkerbox.getName());
+                    tileentityshulkerbox.setCustomName("");
+                }
+
+                spawnAsEntity(worldIn, pos, itemstack);
+            }
+
+            worldIn.updateComparatorOutputLevel(pos, state.getBlock());
+        }
     }
+    // CraftBukkit end
 
     public void onBlockPlacedBy(World worldIn, BlockPos pos, IBlockState state, EntityLivingBase placer, ItemStack stack)
     {
@@ -181,7 +204,7 @@
     {
         TileEntity tileentity = worldIn.getTileEntity(pos);
 
-        if (tileentity instanceof TileEntityShulkerBox)
+        if (false && tileentity instanceof TileEntityShulkerBox) // CraftBukkit - moved up
         {
             TileEntityShulkerBox tileentityshulkerbox = (TileEntityShulkerBox)tileentity;
 
@@ -202,8 +225,9 @@
                 spawnAsEntity(worldIn, pos, itemstack);
             }
 
-            worldIn.updateComparatorOutputLevel(pos, state.getBlock());
+//            worldIn.updateComparatorOutputLevel(pos, state.getBlock());
         }
+        worldIn.updateComparatorOutputLevel(pos, state.getBlock()); // CraftBukkit - moved down
 
         super.breakBlock(worldIn, pos, state);
     }
