--- ../src-base/minecraft/net/minecraftforge/oredict/ShapelessOreRecipe.java
+++ ../src-work/minecraft/net/minecraftforge/oredict/ShapelessOreRecipe.java
@@ -42,6 +42,7 @@
 import com.google.gson.JsonElement;
 import com.google.gson.JsonObject;
 import com.google.gson.JsonParseException;
+import org.bukkit.inventory.Recipe;
 
 public class ShapelessOreRecipe extends IForgeRegistryEntry.Impl<IRecipe> implements IRecipe
 {
@@ -138,9 +139,9 @@
     }
 
     @Override
-    public boolean canFit(int p_194133_1_, int p_194133_2_)
+    public boolean canFit(int width, int height)
     {
-        return p_194133_1_ * p_194133_2_ >= this.input.size();
+        return width * height >= this.input.size();
     }
 
     public static ShapelessOreRecipe factory(JsonContext context, JsonObject json)
@@ -157,4 +158,15 @@
         ItemStack itemstack = CraftingHelper.getItemStack(JsonUtils.getJsonObject(json, "result"), context);
         return new ShapelessOreRecipe(group.isEmpty() ? null : new ResourceLocation(group), ings, itemstack);
     }
+
+
+    @Override
+    public Recipe toBukkitRecipe() {
+        throw new UnsupportedOperationException("Cannot get bukkit-recipe from ShapelessOreRecipe!");
+    }
+
+    @Override
+    public void setKey(ResourceLocation key) {
+
+    }
 }
