--- ../src-base/minecraft/net/minecraft/item/crafting/IRecipe.java
+++ ../src-work/minecraft/net/minecraft/item/crafting/IRecipe.java
@@ -3,6 +3,7 @@
 import net.minecraft.inventory.InventoryCrafting;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.NonNullList;
+import net.minecraft.util.ResourceLocation;
 import net.minecraft.world.World;
 
 public interface IRecipe extends net.minecraftforge.registries.IForgeRegistryEntry<IRecipe>
@@ -34,4 +35,14 @@
     {
         return "";
     }
+
+  default org.bukkit.inventory.Recipe toBukkitRecipe() // CraftBukkit
+  {
+    return null;
+  }
+
+  default void setKey(ResourceLocation key) // CraftBukkit
+  {
+
+  }
 }
