--- ../src-base/minecraft/net/minecraft/client/gui/GuiIngame.java
+++ ../src-work/minecraft/net/minecraft/client/gui/GuiIngame.java
@@ -499,7 +499,7 @@
                     // FORGE - Move status icon check down from above so renderHUDEffect will still be called without a status icon
                     if (potion.hasStatusIcon())
                     this.drawTexturedModalRect(k + 3, l + 3, i1 % 8 * 18, 198 + i1 / 8 * 18, 18, 18);
-                    potion.renderHUDEffect(potioneffect, this, k, l, this.zLevel, f);
+                    potion.renderHUDEffect(k, l, potioneffect, mc, f);
                 }
             }
         }
