--- ../src-base/minecraft/net/minecraftforge/common/ForgeInternalHandler.java
+++ ../src-work/minecraft/net/minecraftforge/common/ForgeInternalHandler.java
@@ -96,7 +96,6 @@
         if (event.phase == Phase.END)
             FMLClientHandler.instance().updateCloudSettings();
     }
-
     @SubscribeEvent
     public void onChunkUnload(ChunkEvent.Unload event)
     {
@@ -104,4 +103,3 @@
             FarmlandWaterManager.removeTickets(event.getChunk());
     }
 }
-
